### ISSUE TITLE
Make sure that permissions exist

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5732]
This came up in various comments in #5294:

@indigoxela 
> @argiepiano have fun with devel and something like `debug(user_access('synchronize whatsoever foobar'));` - This is always true for user 1. ... The "fun" is that user 1 is allowed to do things that don't even exist. 😏 Isn't it cool to be so powerful?

Indeed, from the code in [`user_access()`](https://docs.backdropcms.org/api/backdrop/1/search/user_access):
```php
  // User #1 has all privileges:
  if ($account->uid == 1) {
    return TRUE;
  }
```

As you see, there is no check to see if the permission exists. A permission could be:
- mistyped
- originating from a disabled/uninstalled module
- be something random/obscure

I would like us to make it so that when access for a specific permission is checked, we ensure that that permission exists - if not, it makes more sense to me that we should be returning `FALSE` in `user_access()` (and be logging something about it).

PS: In Drupal-land, this related change has happened in 9.3.x: [Permissions must exist](https://www.drupal.org/node/3193348)

>Starting with Drupal 9.3.0, all permissions in a user role must be defined in a module.permissions.yml file or a permissions callback. Other permissions are now considered invalid. This includes permissions from uninstalled modules, permissions that depend on configuration that has been removed (such as a content type), and permissions of obscure origin.

**Comment by klonos** [comment]
...to be clear, I'm not suggesting that we should be necessarily cleaning up user.role.*.json files from invalid permissions (which is what drupal does) - just make sure that a permission passed in `user_access()` actually exists.

**Comment by klonos** [comment]
...been trying this on my local. Code changes are fairly minimal, but currently hitting memory allocation issues (even after trying to use `backdrop_static_fast`). Pausing for now, but I'll push my code in a PR whether I manage to figure out these memory issues or not.

**Comment by quicksketch** [comment]
Sounds like a great idea :+1:

**Comment by kiamlaluno** [comment]
I take the code should be changed to something similar to the following one.

```php
function user_access($permission_name, $account = NULL) {
  global $user;

  if (!isset($account)) {
    $account = $user;
  }

  // User #1 has all privileges:
  if ($account->uid == 1) {
    // Verify the permission has been defined by a module.
    $permissions = user_permission_get_modules();
    return isset($permissions[$permission_name]);
  }

  // To reduce the number of SQL queries, we cache the user's permissions
  // in a static variable.
  // Use the advanced backdrop_static() pattern, since this is called very often.
  static $backdrop_static_fast;
  if (!isset($backdrop_static_fast)) {
    $backdrop_static_fast['perm'] = &backdrop_static(__FUNCTION__);
  }
  $perm = &$backdrop_static_fast['perm'];
  if (!isset($perm[$account->uid])) {
    $perm[$account->uid] = user_role_permissions($account->roles);
  }

  return in_array($permission_name, $perm[$account->uid]);
}
```

Should the value returned by [`user_permission_get_modules()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21user%21user.module/function/user_permission_get_modules/1) be cached, in some way?

**Comment by kiamlaluno** [comment]
Actually, instead of using `user_permission_get_modules()`, which returns all the permissions, the code should call each `hook_permission()` implementation and verify the returned permissions contain `$permission_name`; if `$permission_name` isn't between those permissions, the code should pass to the next `hook_permission()` implementation until `$permission_name` is found or the `hook_permission()` implementations are all invoked.

`backdrop_static()` would fill the memory with all the permissions. I would rather use a cache, but in that case, code to clear that cache should be added. That cache could be used from `user_permission_get_modules()` too, though.

**Comment by klonos** [comment]
> Actually, instead of using `user_permission_get_modules()`, which returns all the permissions, the code should invoke each `hook_permission()` implementation and verify the returned permissions contain `$permission_name`; ...

In the PR I've filed for #4097, I am bailing out of the loop that looks for implementations of `hook_permission()` if the specific permission is found early - no need to keep looking.

**Comment by klonos** [comment]
...so if the other PR is accepted, then we could simply do this here:

```php
  // User #1 has all privileges:
  if ($account->uid == 1 && user_permission_get_info($permission_name)) {
    return TRUE;
  }
```

...or we could even add a specific `user_permission_exists()` function for it.

**Comment by kiamlaluno** [comment]
@klonos Right, I forgot about that issue. It's better to see what happens in that issue, then. 😉
